migrations/20240210165254_variants__reference_course_id.sql

```diff
@@ -0,0 +1,13 @@
+ALTER TABLE variants
+ADD COLUMN IF NOT EXISTS course_id bigint;
+
+ALTER TABLE variants
+ADD CONSTRAINT variants_course_id_fkey FOREIGN KEY (course_id) REFERENCES pl_courses (id) ON UPDATE CASCADE ON DELETE CASCADE;
+
+UPDATE variants AS v
+SET
+  course_id = q.course_id
+FROM
+  questions AS q
+WHERE
+  v.question_id = q.id;
```

Comment:
@mwest1066 is there any way to do this that isn't going to completely lock the variants table and thus take down PL? Or is there no way around it and we're just going to have to deploy during a low/no usage time?

Reply:
I'll either need to manually run this in chunks, or (much preferred by me) hold off on merging this until I can implement something like this: https://docs.gitlab.com/ee/development/database/batched_background_migrations.html.

We'll want to add a `NOT NULL` constraint once this has been deployed. We can use the method described here to add it without locking the entire table: https://dba.stackexchange.com/a/268128

Reply:
I'm happy to make another PR to add the `NOT NULL` constraint, to be deployed after this one.

Reply:
I've started working on infrastructure for automatically executing batched migrations: #7292. It borrows the basic idea of GitLab's system (a function that is repeatedly invoked with different ranges of a table's ID space), but without quite as much of the complexity.

Reply:
Cool! Feel free to push directly to this PR branch if you want to use it for testing out that infra. Or I'm happy to work on it with you. just let me know.

Reply:
> We'll want to add a `NOT NULL` constraint once this has been deployed. We can use the method described here to add it without locking the entire table: https://dba.stackexchange.com/a/268128

Oh wow, I hadn't known that PG 12 added this way to add a NOT NULL without a full table lock and scan. So awesome!

It would be great to use this to add a lot more NOT NULL constraints to our tables. We have many places (especially for foreign keys) where we are assuming that columns are not null but we don't actually have a NOT NULL constraint in place. I was better about this for PT but in the original PL schema I hadn't appreciated how useful it is to to add NOT NULL. This will be especially helpful when we add zod types corresponding to the DB schema.

Reply:
Ya, it looks pretty awesome. Open some issues for the other ones you want to do?

Reply:
https://github.com/PrairieLearn/PrairieLearn/issues/7320